backend/src/events/controllers.ts

```diff
@@ -39,13 +45,68 @@ const deleteEvent = async (eventID: string) => {
 };
 
 /**
- * Get all events in DB
- * @returns promise with all events or error
+ * Search, Sort and Pagination for Events
+ * [option] corresponds to the columns in the Event table.
+ * Search supports multiple queries
+ * The following options are supported:
+ * @param upcoming if the event is upcoming then true, else false
+ * @param sort possible sort columns: name, location
+ * @param ownerid all events that were created by ownerid
+ * @param userid all events such that userid is registered 
+ * @param req: Request paramters to get query used for search
+ * @returns  promise with list of all events where [option] is [value]
+
  */
-const getEvents = async () => {
-  return prisma.event.findMany({});
-};
+const getEvents = async (req: Request) => {
+  const query = req.query;
+  let upcoming = query.upcoming;
+  let whereDict: { [key: string]: any } = {};
+  let includeDict: { [key: string]: any } = {};
+  if (upcoming) {
+    if (upcoming == "true") {
+      const dateTime = new Date();
+      whereDict["startDate"] = {
+        gt: dateTime,
+      };
+    }
+  }
+  const sortQuery = query.sort as string;
+  const defaultCursor = { id: "asc" };
+  const querySplit = sortQuery ? sortQuery.split(":") : ["default", "asc"];
+  const key: string = querySplit[0];
+  const order = querySplit[1] as Prisma.SortOrder;
+  const sortDict: { [key: string]: any } = {
+    default: { id: order },
+    name: [{ name: order }, defaultCursor],
+    location: [{ location: order }, defaultCursor],
+  };
+
+  const ownerId = query.ownerid;
+  if (ownerId) {
+    whereDict["ownerId"] = ownerId;
+  }
+  const userId = query.userid;
+  if (userId) {
+    whereDict["attendees"] = {
+      some: {
+        userId: userId,
+      },
+    };
+    includeDict["attendees"] = true;
+  }
 
+  return prisma.event.findMany({
+    where: {
+      AND: [whereDict],
+    },
+    include: includeDict,
+    orderBy: sortDict[key],
+    take: query.limit ? parseInt(query.limit as string) : 10,
+    cursor: {
+      id: query.after ? (query.after as string) : "",
+    },
```

Comment:
I have questions on how cursor works.  If you take look at this link https://www.prisma.io/docs/concepts/components/prisma-client/pagination#:~:text=%7D)-,Cursor%2Dbased%20pagination,-Cursor%2Dbased%20pagination
In the code, I think yours is missing the part where they bookmark the location in the result set. Then how does cursor based pagination help us because the cursor is never updated

Reply:
Making cursor required will not be ideal. For example, I copied the ID of the user just after Alice in prisma studio in this query and I didn't get any result. 
`users?firstName=Alice&sort=email:asc&limit=100&after=clo0s0ax40001ccb712nntgbf`

Reply:
> I have questions on how cursor works. If you take look at this link https://www.prisma.io/docs/concepts/components/prisma-client/pagination#:~:text=%7D)-,Cursor%2Dbased%20pagination,-Cursor%2Dbased%20pagination In the code, I think yours is missing the part where they bookmark the location in the result set. Then how does cursor based pagination help us because the cursor is never updated

I thought that it was used only for demontration, since `myCursor` is only a variable and not necessarily part of the implementation. I could however, return the response containing the `cursor id`. 

> Making cursor required will not be ideal. For example, I copied the ID of the user just after Alice in prisma studio in this query and I didn't get any result.
users?firstName=Alice&sort=email:asc&limit=100&after=clo0s0ax40001ccb712nntgbf

I will fix this right now.

Reply:
Hi Akin, all changes have been applied. Hopefully it is to your expectation. :)
- For retrieving the first page, `cursor id` can be omitted. Next cursor is incorporated in the response.
<img width="931" alt="image" src="https://github.com/cornellh4i/lagos-volunteers/assets/68068854/9204494f-f412-49ce-9c0e-2bf493c5e02d">

- For future retrievals, `cursor id` is needed.
<img width="838" alt="image" src="https://github.com/cornellh4i/lagos-volunteers/assets/68068854/820e4387-efe5-4a9b-a55d-efc6c2000777">

Reply:
Hi, Yichen. This looks really clever. I will test and revert

Reply:
Hi Akin, I looked into the database and currently we only have 1 enrollment. Did you add more attendees manually for this?

Reply:
Hey, yes! I did. I modified my seed scripts which populated my database with a lot user enrollment. Can you copy the entire seed file from here - https://github.com/cornellh4i/lagos-volunteers/blob/event-registration-fullstack/backend/prisma/seed.ts 

You will also need to update `backend/src/utils/script.ts` https://github.com/cornellh4i/lagos-volunteers/blob/event-registration-fullstack/backend/src/utils/script.ts